### PR TITLE
SE-0036 (3/3): Provide errors when referencing enum elements on instances

### PR DESCRIPTION
**Pull request by ahoppen**
#### What's in this pull request?

This is part 3/3 implementing [SE-0036](https://github.com/apple/swift-evolution/blob/master/proposals/0036-enum-dot.md).

It undos the changes done in #2842 that provided warnings when an enum element was referenced on an enum and implements a different logic that throws errors instead.

It has slightly better fix-its than the warning-based approach because it can also detect if a single dot suffices as a fix-it in expressions where the type can be inferred on the context, e.g. in `self = .A`. However, it looses the ability to infer a single-dot fix-it in `case` statements for generic enums. 

The resulting code is basically the same as in #2634. 
##### Dependencies

Same text as in #2842:
I wasn’t sure how I should handle the multi-staged rollout of SE-0036 as described [here](https://github.com/apple/swift/pull/2634#issuecomment-221028951). So this PR includes the changes to the test cases that shall be merged in #2841 and #2842 and builds on top of them, i.e. the branch for this PR is two commits ahead of the one for #2842. I hope that’s the right way to go, otherwise let me know.
#### Resolved bug number: ([SR-1236](https://bugs.swift.org/browse/SR-1236))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -139,6 +139,8 @@ Swift 3.0
     `NS[Mutable]Dictionary` are still imported as nongeneric classes for
     the time being.
 
+* [SE-0036](https://github.com/apple/swift-evolution/blob/master/proposals/0036-enum-dot.md): 
+  Enum elements can no longer be accessed like instance members in instance methods.
```
Comment (slavapestov) [inline CHANGELOG.md]: Please add a newline after for consistency

**Comment by slavapestov** [comment]
@DougGregor or @lattner can you take a look at this patch?

File: CHANGELOG.md (hunk 1 above)
Comment (ahoppen) [inline CHANGELOG.md]: Done

**Comment by ahoppen** [comment]
I guess this will still need to be merged for Swift 3

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by ahoppen** [comment]
Reverting the changes for the warnings fixed a crash.
The other failure was fixed in 2c10bec3.

@swift-ci please test and merge

**Comment by DougGregor** [comment]
The LLDB failures are real... it looks like there are a few test cases that need to be updated there. @ahoppen, can you create a PR against LLDB to fix up those errors? We'll merge that one first.

**Comment by ahoppen** [comment]
LLDB pull request is created https://github.com/apple/swift-lldb/pull/35

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
LLDB PR merged, so I kicked off another round of smoke-test-and-merge

**Pull request merged**